modin/core/dataframe/pandas/dataframe/dataframe.py

```diff
@@ -2621,6 +2631,8 @@ def broadcast_apply_full_axis(
         other,
         new_index=None,
         new_columns=None,
+        new_row_lengths=None,
+        new_column_widths=None,
```

Comment:
I'm not sure about these parameters and the fact that they're present in a public method. Using them may require a deep understanding of internal partitioning logic (e.g. dealing with empty partitions) that actually shouldn't be disclosed to the outer world.

My concern is that it's very easy to break everything with these parameters if you don't fully understand what's happening underneath. Here's an example of how easy it is to end up with an inconsistent frame using these parameters:

```python
import modin.pandas as pd
import modin.config as cfg
cfg.NPartitions.put(4)

df = pd.DataFrame({"a": [1, 2, 3], "b": [3, 4, 5]})
qc = df._query_compiler

def make_a_series(df):
    series = df[df.columns[0]]
    res = series.to_frame()
    assert len(res.columns) == 1
    return res

mf = qc._modin_frame.apply_full_axis(
    axis=1,
    # Completely legitimate arguments from high-level perspective
    func=make_a_series,
    new_columns=["__reduced__"],
    new_column_widths=[1],
    keep_partitioning=False
)

print(mf.column_widths) # [1], when it should be: [1, 0, 0, 0]
print(mf._partitions.shape) # (1, 4)
```

I would propose instead the cache to be computed inside the `.broadcast_apply_full_axis`:
```python
if new_index is not None:
    if new_partitions.shape[0] == 1:
        kw["row_lengths"] = [len(new_index)]
    # TODO: add more cases where we can pre-compute partition shapes
if new_columns is not None:
    if new_partitions.shape[1] == 1:
        kw["column_widths"] = [len(new_columns)]
    # TODO: add more cases where we can pre-compute partition shapes
```

Reply:
Just came to my mind that using `self._modin_frame._partitions.shape[1] > 1:` condition may involve an unnecessary step (when there are empty partitions except one non-empty), i.e. performing `apply_full_axis` over axis=1 and then axis=0. I propose to leave `if self.columns > 1:` condition and speed up the things step by step later if we will feel that is necessary. So reverting the changes.

Reply:
Reverted. Let's push this forward.

Reply:
we could just leave the combined condition then...
```python
if self._modin_frame._partitions.shape[1] > 1 and self.columns > 1:
```
okay, let's do it in a separate pr

Reply:
> okay, let's do it in a separate pr

@dchigarev do we issue or smth for this?

Reply:
> do we issue or smth for this?

@Garra1980 
yes, it's #5613.

We also have an epic for `drop_duplicates` optimizations where store all related tasks: #5612

Reply:
right, thanks!